File: backend/app/api/runtime_enterprise.py
"""
企业级 Agent Runtime API

提供:
    1. 任务提交 / 查询 / 取消 / 重试
    2. SSE 实时事件流
    3. WebSocket 双向通信
    4. Dispatcher / WorkerPool / Scheduler 统计

架构:
    API Layer → TaskDispatcher → WorkerPool → AgentWorker → Agent → ResponseCollector
"""
import asyncio
import json
import logging
from typing import Any, Dict, Optional

from fastapi import APIRouter, WebSocket, WebSocketDisconnect, HTTPException, Query, Depends
from pydantic import BaseModel, Field
from sse_starlette.sse import EventSourceResponse

from app.core.logger import log
from app.api.auth import require_auth
from app.models.user import User

router = APIRouter()
logger = logging.getLogger(__name__)


# ============================================================
# 请求 / 响应模型
# ============================================================

class SubmitTaskRequest(BaseModel):
    """提交任务请求"""
    agent_name: str = Field(..., description="目标 Agent 名称")
    action: str = Field("execute", description="调用的 action")
    payload: Dict[str, Any] = Field(default_factory=dict, description="任务参数")
    priority: str = Field("normal", description="优先级: low/normal/high/urgent")
    timeout_seconds: int = Field(300, description="超时时间(秒)")
    max_retries: int = Field(3, description="最大重试次数")
    task_type: str = Field("agent", description="任务类型: agent/flow")
    wait: bool = Field(False, description="是否同步等待结果")


class ScheduleTaskRequest(BaseModel):
    """调度任务请求"""
    agent_name: str = Field(..., description="目标 Agent 名称")
    action: str = Field("execute", description="调用的 action")
    payload: Dict[str, Any] = Field(default_factory=dict, description="任务参数")
    delay_seconds: float = Field(0, description="延迟秒数(0=立即)")
    cron_expression: str = Field("", description="cron 表达式(定时)")
    max_runs: Optional[int] = Field(None, description="最大执行次数(None=无限)")


class TaskResponse(BaseModel):
    """任务响应"""
    code: int = 0
    message: str = "success"
    data: Optional[Any] = None


# ============================================================
# 任务管理接口
# ============================================================

@router.post("/tasks", summary="提交任务", response_model=TaskResponse)
async def submit_task(
    req: SubmitTaskRequest,
    user: User = Depends(require_auth),
):
    """提交任务到 TaskDispatcher

    - `wait=false` (默认): 立即返回 task_id,异步执行
    - `wait=true`: 同步等待结果后返回
    """
    from app.runtime.enterprise import (
        get_task_dispatcher, TaskRequest, TaskPriority,
    )

    try:
        dispatcher = get_task_dispatcher()
        priority = TaskPriority(req.priority)
        request = TaskRequest(
            agent_name=req.agent_name,
            action=req.action,
            payload=req.payload,
            priority=priority,
            user_id=user.id,
            timeout_seconds=req.timeout_seconds,
            max_retries=req.max_retries,
            task_type=req.task_type,
        )

        if req.wait:
            # 同步等待
            result = await dispatcher.submit_and_wait(request)
            return TaskResponse(data=result.to_dict())
        else:
            # 异步提交
            task_id = await dispatcher.submit(request)
            return TaskResponse(data={"task_id": task_id})
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.error(f"提交任务失败: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail=f"提交失败: {e}")


@router.get("/tasks/{task_id}", summary="查询任务状态")
async def get_task_status(
    task_id: str,
    user: User = Depends(require_auth),
):
    """查询任务状态"""
    from app.runtime.enterprise import get_task_dispatcher

    dispatcher = get_task_dispatcher()
    state = await dispatcher.get_status(task_id)
    if state is None:
        raise HTTPException(status_code=404, detail=f"任务不存在: {task_id}")
    return {"code": 0, "data": state.to_dict()}


@router.get("/tasks", summary="列出任务")
async def list_tasks(
    status: Optional[str] = Query(None, description="过滤状态"),
    limit: int = Query(50, ge=1, le=500),
    user: User = Depends(require_auth),
):
    """列出任务"""
    from app.runtime.enterprise import get_task_dispatcher, TaskStatus

    dispatcher = get_task_dispatcher()
    task_status = TaskStatus(status) if status else None
    tasks = await dispatcher.list_tasks(status=task_status, limit=limit)
    return {"code": 0, "data": [t.to_dict() for t in tasks]}


@router.get("/tasks/{task_id}/result", summary="获取任务结果")
async def get_task_result(
    task_id: str,
    user: User = Depends(require_auth),
):
    """获取已完成任务的结果"""
    from app.runtime.enterprise import get_task_dispatcher

    dispatcher = get_task_dispatcher()
    result = await dispatcher.get_result(task_id)
    if result is None:
        raise HTTPException(status_code=404, detail=f"结果不存在: {task_id}")
    return {"code": 0, "data": result.to_dict()}


@router.post("/tasks/{task_id}/cancel", summary="取消任务")
async def cancel_task(
    task_id: str,
    user: User = Depends(require_auth),
):
    """取消任务"""
    from app.runtime.enterprise import get_task_dispatcher

    dispatcher = get_task_dispatcher()
    success = await dispatcher.cancel(task_id)
    if not success:
        raise HTTPException(status_code=400, detail="任务不存在或已终态")
    return {"code": 0, "message": "已取消"}


@router.post("/tasks/{task_id}/retry", summary="重试任务")
async def retry_task(
    task_id: str,
    user: User = Depends(require_auth),
):
    """重试失败的任务"""
    from app.runtime.enterprise import get_task_dispatcher

    dispatcher = get_task_dispatcher()
    success = await dispatcher.retry(task_id)
    if not success:
        raise HTTPException(status_code=400, detail="任务不可重试")
    return {"code": 0, "message": "重试已启动"}


# ============================================================
# SSE 实时事件流
# ============================================================

@router.get("/stream/{task_id}", summary="SSE 事件流")
async def sse_stream(
    task_id: str,
    # user: User = Depends(require_auth),  # SSE 不方便带 token,暂不鉴权
):
    """SSE 实时事件流

    前端使用 EventSource 订阅:
        const es = new EventSource('/api/runtime/stream/task_xxx');
        es.onmessage = (e) => console.log(JSON.parse(e.data));
        es.addEventListener('done', () => es.close());

    事件类型:
        - start:    任务开始
        - progress:  进度更新
        - end:       任务完成
        - error:     错误
        - done:      流结束 (信号)
        - ping:      心跳保活
    """
    from app.runtime.enterprise import get_stream_publisher

    publisher = get_stream_publisher()

    async def event_generator():
        try:
            async for event in publisher.sse_stream(task_id):
                # SSE 格式: event + data
                event_type = event.get("event", "message")
                yield {
                    "event": event_type,
                    "data": json.dumps(event, ensure_ascii=False, default=str),
                }
                if event_type == "done":
                    break
        except asyncio.CancelledError:
            logger.info(f"SSE 流被取消: {task_id}")
        except Exception as e:
            logger.error(f"SSE 流异常: {task_id} | {e}", exc_info=True)
            yield {
                "event": "error",
                "data": json.dumps({"error": str(e), "task_id": task_id}),
            }

    return EventSourceResponse(event_generator())


# ============================================================
# WebSocket 双向通信
# ============================================================

@router.websocket("/ws/{task_id}")
async def websocket_endpoint(websocket: WebSocket, task_id: str):
    """WebSocket 双向通信

    前端使用:
        const ws = new WebSocket('ws://localhost:8000/api/runtime/ws/task_xxx');
        ws.onmessage = (e) => console.log(JSON.parse(e.data));
        ws.onclose = () => console.log('连接关闭');

    客户端可发送:
        - "ping": 心跳,服务端返回 "pong"
        - "cancel": 取消订阅
    """
    from app.runtime.enterprise import get_stream_publisher

    await websocket.accept()
    publisher = get_stream_publisher()

    await publisher.add_ws_client(task_id, websocket)
    logger.info(f"WebSocket 连接 | task_id={task_id}")

    try:
        # 启动服务循环 (接收客户端消息)
        await publisher.serve_ws(task_id, websocket)
    except WebSocketDisconnect:
        logger.info(f"WebSocket 断开 | task_id={task_id}")
    except Exception as e:
        logger.error(f"WebSocket 异常: {task_id} | {e}", exc_info=True)
    finally:
        await publisher.remove_ws_client(task_id, websocket)


# ============================================================
# 调度接口
# ============================================================

@router.post("/schedule", summary="调度任务")
async def schedule_task(
    req: ScheduleTaskRequest,
    user: User = Depends(require_auth),
):
    """调度任务 (延迟 / 定时)"""
    from app.runtime.enterprise import (
        get_task_scheduler, get_task_dispatcher, TaskRequest,
    )

    try:
        dispatcher = get_task_dispatcher()
        scheduler = get_task_scheduler(dispatcher=dispatcher)

        request = TaskRequest(
            agent_name=req.agent_name,
            action=req.action,
            payload=req.payload,
            user_id=user.id,
        )

        if req.cron_expression:
            sched_id = await scheduler.schedule_cron(
                request=request,
                cron_expression=req.cron_expression,
                max_runs=req.max_runs,
            )
            return {"code": 0, "data": {"schedule_id": sched_id, "type": "cron"}}
        else:
            sched_id = await scheduler.schedule(
                request=request,
                delay_seconds=req.delay_seconds,
            )
            return {"code": 0, "data": {"schedule_id": sched_id, "type": "delayed"}}
    except Exception as e:
        logger.error(f"调度失败: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/schedules", summary="列出调度任务")
async def list_schedules(
    enabled_only: bool = Query(False),
    user: User = Depends(require_auth),
):
    """列出所有调度任务"""
    from app.runtime.enterprise import get_task_scheduler

    scheduler = get_task_scheduler()
    schedules = await scheduler.list_schedules(enabled_only=enabled_only)
    return {"code": 0, "data": [s.to_dict() for s in schedules]}


@router.delete("/schedules/{schedule_id}", summary="取消调度")
async def cancel_schedule(
    schedule_id: str,
    user: User = Depends(require_auth),
):
    """取消调度任务"""
    from app.runtime.enterprise import get_task_scheduler

    scheduler = get_task_scheduler()
    success = await scheduler.cancel(schedule_id)
    if not success:
        raise HTTPException(status_code=404, detail="调度任务不存在")
    return {"code": 0, "message": "已取消"}


# ============================================================
# 统计接口
# ============================================================

@router.get("/stats", summary="Runtime 统计")
async def get_runtime_stats(
    user: User = Depends(require_auth),
):
    """获取 Runtime 全部统计信息"""
    from app.runtime.enterprise import (
        get_task_dispatcher, get_worker_pool,
        get_response_collector, get_stream_publisher,
        get_task_scheduler,
    )

    dispatcher = get_task_dispatcher()
    pool = get_worker_pool()
    collector = get_response_collector()
    publisher = get_stream_publisher()
    scheduler = get_task_scheduler()

    return {
        "code": 0,
        "data": {
            "dispatcher": await dispatcher.get_stats(),
            "worker_pool": await pool.get_stats(),
            "collector": await collector.get_stats(),
            "stream_publisher": await publisher.get_stats(),
            "scheduler": await scheduler.get_stats(),
        },
    }


@router.get("/workers", summary="Worker 统计")
async def get_worker_stats(
    user: User = Depends(require_auth),
):
    """获取 WorkerPool 统计"""
    from app.runtime.enterprise import get_worker_pool

    pool = get_worker_pool()
    return {"code": 0, "data": await pool.get_stats()}


@router.get("/health", summary="Runtime 健康检查")
async def runtime_health():
    """Runtime 健康检查 (无需鉴权)"""
    from app.runtime.enterprise import get_task_dispatcher

    try:
        dispatcher = get_task_dispatcher()
        stats = await dispatcher.get_stats()
        return {
            "status": "healthy",
            "mode": stats["mode"],
            "started": stats["started"],
            "queue_size": stats["queue_size"],
        }
    except Exception as e:
        return {"status": "unhealthy", "error": str(e)}


# ============================================================
# 历史任务查询
# ============================================================

@router.get("/history", summary="查询历史任务")
async def query_history(
    status: Optional[str] = Query(None, description="过滤状态"),
    agent_name: Optional[str] = Query(None, description="过滤 Agent"),
    hours: int = Query(24, ge=1, le=720, description="查询最近 N 小时"),
    limit: int = Query(100, ge=1, le=500),
    offset: int = Query(0, ge=0),
    user: User = Depends(require_auth),
):
    """查询历史任务 (从数据库)

    支持按状态、Agent、时间范围过滤。
    """
    from app.services.runtime_persistence import get_runtime_persistence
    from datetime import datetime, timedelta

    persistence = get_runtime_persistence()
    if not persistence.enabled:
        raise HTTPException(status_code=503, detail="持久化服务未启用")

    start_time = datetime.now() - timedelta(hours=hours)
    tasks, total = persistence.query_history(
        status=status,
        agent_name=agent_name,
        user_id=user.id,
        start_time=start_time,
        limit=limit,
        offset=offset,
    )
    return {
        "code": 0,
        "data": {
            "tasks": tasks,
            "total": total,
            "limit": limit,
            "offset": offset,
        },
    }


@router.get("/history/{task_id}", summary="查询历史任务详情")
async def get_history_task(
    task_id: str,
    user: User = Depends(require_auth),
):
    """按 task_id 查询历史任务详情"""
    from app.services.runtime_persistence import get_runtime_persistence

    persistence = get_runtime_persistence()
    if not persistence.enabled:
        raise HTTPException(status_code=503, detail="持久化服务未启用")

    task = persistence.get_task_by_id(task_id)
    if task is None:
        raise HTTPException(status_code=404, detail=f"任务不存在: {task_id}")
    return {"code": 0, "data": task}


# ============================================================
# 指标统计
# ============================================================

@router.get("/metrics", summary="Runtime 指标统计")
async def get_runtime_metrics(
    hours: int = Query(24, ge=1, le=720, description="统计最近 N 小时"),
    agent_name: Optional[str] = Query(None, description="按 Agent 过滤"),
    user: User = Depends(require_auth),
):
    """获取 Runtime 指标统计

    返回:
        - total: 总任务数
        - by_status: 按状态分组
        - success_rate: 成功率
        - avg_duration_ms: 平均耗时
        - p50_duration_ms: 中位数耗时
        - p95_duration_ms: 95 分位耗时
        - throughput: 吞吐量 (任务/分钟)
        - by_agent: 按 Agent 分组
        - hourly: 按小时分组的趋势
    """
    from app.services.runtime_persistence import get_runtime_persistence

    persistence = get_runtime_persistence()
    if not persistence.enabled:
        raise HTTPException(status_code=503, detail="持久化服务未启用")

    metrics = persistence.get_metrics(hours=hours, agent_name=agent_name)
    return {"code": 0, "data": metrics}

File: frontend/src/services/runtimeEnterprise.ts
/**
 * 企业级 Agent Runtime 服务
 *
 * 后端路由前缀: /runtime
 * 接口分组:
 *   1. 任务管理  - POST /tasks, GET /tasks, GET /tasks/{id}, ...
 *   2. SSE 流    - GET /stream/{task_id}
 *   3. WebSocket - WS /ws/{task_id}
 *   4. 调度管理  - POST /schedule, GET /schedules
 *   5. 统计信息  - GET /stats, GET /workers, GET /health
 */
import request from './request';

// ============================================================
// 类型定义
// ============================================================

export type TaskStatus = 'pending' | 'running' | 'success' | 'failed' | 'timeout' | 'cancelled';
export type TaskPriority = 'low' | 'normal' | 'high' | 'urgent';
export type DispatcherMode = 'standalone' | 'distributed';

export interface TaskState {
  task_id: string;
  task_type: string;
  agent_name: string;
  action: string;
  payload: Record<string, any>;
  status: TaskStatus;
  priority: TaskPriority;
  retry_count: number;
  max_retries: number;
  created_at: string;
  started_at?: string | null;
  completed_at?: string | null;
  result?: any;
  error?: string | null;
  events_count: number;
  user_id?: number | null;
  session_id?: string | null;
  timeout_seconds: number;
  worker_id?: string | null;
}

export interface TaskResult {
  task_id: string;
  status: string;
  result?: any;
  error?: string | null;
  events_count: number;
  duration_ms: number;
  worker_id?: string | null;
  retry_count: number;
}

export interface SubmitTaskRequest {
  agent_name: string;
  action?: string;
  payload?: Record<string, any>;
  priority?: TaskPriority;
  timeout_seconds?: number;
  max_retries?: number;
  task_type?: string;
  wait?: boolean;
}

export interface ScheduleTaskRequest {
  agent_name: string;
  action?: string;
  payload?: Record<string, any>;
  delay_seconds?: number;
  cron_expression?: string;
  max_runs?: number | null;
}

export interface ScheduledTask {
  schedule_id: string;
  agent_name: string;
  action: string;
  schedule_type: 'immediate' | 'delayed' | 'scheduled' | 'dependent';
  delay_seconds: number;
  cron_expression: string;
  after_task_ids: string[];
  next_run_at?: number | null;
  last_run_at?: number | null;
  run_count: number;
  max_runs?: number | null;
  enabled: boolean;
}

export interface WorkerStats {
  worker_id: string;
  is_busy: boolean;
  current_task?: string | null;
  tasks_done: number;
  tasks_failed: number;
  total_duration_ms: number;
  avg_duration_ms: number;
}

export interface WorkerPoolStats {
  size: number;
  started: boolean;
  busy_workers: number;
  idle_workers: number;
  running_tasks: number;
  workers: WorkerStats[];
}

export interface DispatcherStats {
  mode: DispatcherMode;
  started: boolean;
  queue_size: number;
  total_tasks: number;
  completed_results: number;
  pending_futures: number;
  status_counts: Record<string, number>;
  worker_pool?: string | null;
}

export interface CollectorStats {
  active_tasks: number;
  completed_tasks: number;
  total_subscribers: number;
  max_history: number;
  stream_publisher?: string | null;
}

export interface StreamPublisherStats {
  events_published: number;
  sse_tasks: number;
  sse_clients_total: number;
  ws_tasks: number;
  ws_clients_total: number;
  sse_clients_history: number;
  ws_clients_history: number;
}

export interface SchedulerStats {
  started: boolean;
  total_schedules: number;
  enabled_schedules: number;
  by_type: Record<string, number>;
  dispatcher?: string | null;
}

export interface RuntimeStats {
  dispatcher: DispatcherStats;
  worker_pool: WorkerPoolStats;
  collector: CollectorStats;
  stream_publisher: StreamPublisherStats;
  scheduler: SchedulerStats;
}

export interface RuntimeHealth {
  status: 'healthy' | 'unhealthy';
  mode?: DispatcherMode;
  started?: boolean;
  queue_size?: number;
  error?: string;
}

export interface SSEEvent {
  event: string;
  task_id?: string;
  [key: string]: any;
}

// ============================================================
// 任务管理
// ============================================================

/** 提交任务 */
export async function submitTask(data: SubmitTaskRequest): Promise<{ task_id: string } | TaskResult> {
  const resp = await request.post('/runtime/tasks', data);
  return resp.data;
}

/** 查询任务状态 */
export async function getTaskStatus(taskId: string): Promise<TaskState> {
  const resp = await request.get(`/runtime/tasks/${taskId}`);
  return resp.data;
}

/** 列出任务 */
export async function listTasks(params?: {
  status?: TaskStatus;
  limit?: number;
}): Promise<TaskState[]> {
  const resp = await request.get('/runtime/tasks', { params });
  return resp.data;
}

/** 获取任务结果 */
export async function getTaskResult(taskId: string): Promise<TaskResult> {
  const resp = await request.get(`/runtime/tasks/${taskId}/result`);
  return resp.data;
}

/** 取消任务 */
export async function cancelTask(taskId: string): Promise<void> {
  await request.post(`/runtime/tasks/${taskId}/cancel`);
}

/** 重试任务 */
export async function retryTask(taskId: string): Promise<void> {
  await request.post(`/runtime/tasks/${taskId}/retry`);
}

// ============================================================
// 调度管理
// ============================================================

/** 调度任务 */
export async function scheduleTask(data: ScheduleTaskRequest): Promise<{ schedule_id: string; type: string }> {
  const resp = await request.post('/runtime/schedule', data);
  return resp.data;
}

/** 列出调度任务 */
export async function listSchedules(enabledOnly?: boolean): Promise<ScheduledTask[]> {
  const resp = await request.get('/runtime/schedules', { params: { enabled_only: enabledOnly } });
  return resp.data;
}

/** 取消调度 */
export async function cancelSchedule(scheduleId: string): Promise<void> {
  await request.delete(`/runtime/schedules/${scheduleId}`);
}

// ============================================================
// 统计与健康
// ============================================================

/** Runtime 健康检查 */
export async function getRuntimeHealth(): Promise<RuntimeHealth> {
  const resp = await request.get('/runtime/health');
  return resp.data;
}

/** Runtime 完整统计 */
export async function getRuntimeStats(): Promise<RuntimeStats> {
  const resp = await request.get('/runtime/stats');
  return resp.data;
}

/** Worker 池统计 */
export async function getWorkerStats(): Promise<WorkerPoolStats> {
  const resp = await request.get('/runtime/workers');
  return resp.data;
}

// ============================================================
// 历史任务查询
// ============================================================

export interface HistoryTask {
  id: number;
  task_id: string;
  parent_task_id?: string | null;
  task_type: string;
  agent_name: string;
  action: string;
  status: TaskStatus;
  priority: TaskPriority;
  retry_count: number;
  max_retries: number;
  created_at?: string | null;
  started_at?: string | null;
  completed_at?: string | null;
  duration_ms?: number | null;
  timeout_seconds: number;
  worker_id?: string | null;
  user_id?: number | null;
  session_id?: string | null;
  error?: string | null;
  events_count: number;
}

export interface HistoryQueryParams {
  status?: TaskStatus;
  agent_name?: string;
  hours?: number;
  limit?: number;
  offset?: number;
}

export interface HistoryQueryResult {
  tasks: HistoryTask[];
  total: number;
  limit: number;
  offset: number;
}

/** 查询历史任务 */
export async function queryHistory(params: HistoryQueryParams): Promise<HistoryQueryResult> {
  const resp = await request.get('/runtime/history', { params });
  return resp.data;
}

/** 查询历史任务详情 */
export async function getHistoryTask(taskId: string): Promise<HistoryTask> {
  const resp = await request.get(`/runtime/history/${taskId}`);
  return resp.data;
}

// ============================================================
// 指标统计
// ============================================================

export interface RuntimeMetrics {
  hours: number;
  total: number;
  by_status: Record<string, number>;
  success_rate: number;
  avg_duration_ms: number;
  p50_duration_ms: number;
  p95_duration_ms: number;
  throughput: number;
  by_agent: Record<string, {
    total: number;
    success: number;
    failed: number;
    total_duration_ms: number;
    success_rate: number;
    avg_duration_ms: number;
  }>;
  hourly: Array<{
    hour: string;
    total: number;
    success: number;
    failed: number;
  }>;
}

/** 获取指标统计 */
export async function getRuntimeMetrics(
  hours?: number,
  agentName?: string,
): Promise<RuntimeMetrics> {
  const resp = await request.get('/runtime/metrics', {
    params: { hours, agent_name: agentName },
  });
  return resp.data;
}

/**
 * 创建 SSE 事件流连接
 *
 * 使用方式:
 *   const close = createSSEStream('task_xxx', (event) => {
 *     console.log('收到事件:', event);
 *   });
 *   // 关闭连接
 *   close();
 */
export function createSSEStream(
  taskId: string,
  onEvent: (event: SSEEvent) => void,
  onError?: (error: Event) => void,
): () => void {
  // 注意: EventSource 不支持自定义 header,使用 cookie 认证
  const url = `/api/runtime/stream/${taskId}`;
  const eventSource = new EventSource(url);

  eventSource.onmessage = (e) => {
    try {
      const data = JSON.parse(e.data);
      onEvent(data);
    } catch {
      onEvent({ event: 'message', raw: e.data, task_id: taskId });
    }
  };

  // 监听特定事件类型
  const eventTypes = ['start', 'progress', 'end', 'error', 'done', 'ping'];
  eventTypes.forEach((type) => {
    eventSource.addEventListener(type, (e: any) => {
      try {
        const data = e.data ? JSON.parse(e.data) : {};
        onEvent({ ...data, event: type, task_id: taskId });
      } catch {
        onEvent({ event: type, task_id: taskId });
      }
      if (type === 'done') {
        eventSource.close();
      }
    });
  });

  eventSource.onerror = (e) => {
    if (onError) onError(e);
    eventSource.close();
  };

  // 返回关闭函数
  return () => {
    eventSource.close();
  };
}

/**
 * 创建 WebSocket 连接
 *
 * 使用方式:
 *   const ws = createWebSocket('task_xxx', (event) => {
 *     console.log('收到事件:', event);
 *   });
 *   // 关闭连接
 *   ws.close();
 */
export function createWebSocket(
  taskId: string,
  onEvent: (event: SSEEvent) => void,
  onError?: (error: Event) => void,
): WebSocket {
  const protocol = window.location.protocol === 'https:' ? 'wss:' : 'ws:';
  const host = window.location.host;
  const url = `${protocol}//${host}/api/runtime/ws/${taskId}`;

  const ws = new WebSocket(url);

  ws.onmessage = (e) => {
    try {
      const data = JSON.parse(e.data);
      onEvent(data);
    } catch {
      onEvent({ event: 'message', raw: e.data, task_id: taskId });
    }
  };

  ws.onerror = (e) => {
    if (onError) onError(e);
  };

  return ws;
}
